tracer/src/Datadog.Trace/Tracer.cs

```diff
@@ -380,23 +380,26 @@ internal SpanContext CreateSpanContext(ISpanContext parent = null, string servic
                 var samplingPriority = DistributedTracer.Instance.GetSamplingPriority();
                 traceContext.SetSamplingPriority(samplingPriority);
 
-                if (traceId == null)
+                if (traceId.IsZero() &&
+                    Activity.ActivityListener.GetCurrentActivity() is Activity.DuckTypes.IW3CActivity { TraceId: { } activityTraceId })
                 {
-                    var activity = Activity.ActivityListener.GetCurrentActivity();
-                    if (activity is Activity.DuckTypes.IW3CActivity w3CActivity)
-                    {
-                        // If the Activity has an ActivityIdFormat.Hierarchical instead of W3C it's TraceId & SpanId will be null
-                        if (w3CActivity.TraceId is { } w3cTraceId)
-                        {
-                            // If there's an existing activity we use the same traceId (converted).
-                            rawTraceId = w3cTraceId;
-                            traceId = Convert.ToUInt64(w3cTraceId.Substring(16), 16);
-                        }
-                    }
+                    // if there's an existing Activity we try to use its TraceId,
+                    // but if Activity.IdFormat is not ActivityIdFormat.W3C, it may be null or unparsable
+                    rawTraceId = activityTraceId;
+                    HexString.TryParseTraceId(activityTraceId, out traceId);
                 }
             }
 
             var finalServiceName = serviceName ?? DefaultServiceName;
+
+            if (traceId.IsZero())
+            {
+                // generate the trace id here using the 128-bit setting
+                // instead of letting the SpanContext generate it in its ctor
```

Comment:
I don't understand why exactly we want to generate the traceId here?

Reply:
Thanks! That helps clear it up

+1 to removing more instances of `SpanContext` 😄